File: apps/crud-be-java-vertx/src/main/java/com/demobejavx/repository/memory/InMemoryAttachmentRepository.java
package com.demobejavx.repository.memory;

import com.demobejavx.domain.model.Attachment;
import com.demobejavx.repository.AttachmentRepository;
import io.vertx.core.Future;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public class InMemoryAttachmentRepository implements AttachmentRepository {

    private final ConcurrentHashMap<String, Attachment> store = new ConcurrentHashMap<>();
    private final AtomicLong idSequence = new AtomicLong(1);

    @Override
    public Future<Attachment> save(Attachment attachment) {
        String id = String.valueOf(idSequence.getAndIncrement());
        Attachment saved = attachment.withId(id);
        store.put(id, saved);
        return Future.succeededFuture(saved);
    }

    @Override
    public Future<Optional<Attachment>> findById(String id) {
        return Future.succeededFuture(Optional.ofNullable(store.get(id)));
    }

    @Override
    public Future<List<Attachment>> findByExpenseId(String expenseId) {
        List<Attachment> result = store.values().stream()
                .filter(a -> a.expenseId().equals(expenseId))
                .toList();
        return Future.succeededFuture(new ArrayList<>(result));
    }

    @Override
    public Future<Boolean> deleteById(String id) {
        boolean removed = store.remove(id) != null;
        return Future.succeededFuture(removed);
    }

    public void reset() {
        store.clear();
        idSequence.set(1);
    }
}

File: apps/demo-be-ktkt/src/main/kotlin/com/organiclever/demoktkt/infrastructure/InMemoryAttachmentRepository.kt
package com.organiclever.demoktkt.infrastructure

import com.organiclever.demoktkt.domain.Attachment
import com.organiclever.demoktkt.infrastructure.repositories.AttachmentRepository
import com.organiclever.demoktkt.infrastructure.repositories.CreateAttachmentRequest
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class InMemoryAttachmentRepository : AttachmentRepository {
  private val store = ConcurrentHashMap<UUID, Attachment>()

  override suspend fun create(request: CreateAttachmentRequest): Attachment {
    val attachment =
      Attachment(
        id = UUID.randomUUID(),
        expenseId = request.expenseId,
        userId = request.userId,
        filename = request.filename,
        contentType = request.contentType,
        sizeBytes = request.sizeBytes,
        storedPath = request.storedPath,
        createdAt = Instant.now(),
      )
    store[attachment.id] = attachment
    return attachment
  }

  override suspend fun findById(id: UUID): Attachment? = store[id]

  override suspend fun findAllByExpense(expenseId: UUID): List<Attachment> =
    store.values.filter { it.expenseId == expenseId }

  override suspend fun delete(id: UUID): Boolean = store.remove(id) != null

  fun clear() {
    store.clear()
  }
}
